dali/pipeline/data/tensor_vector.cc

```diff
@@ -0,0 +1,439 @@
+  }
+  return result;
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::Resize(const TensorListShape<> &new_shape, const TypeInfo &new_type) {
+  resize_tensors(new_shape.num_samples());
+  if (state_ == State::contiguous) {
+    tl_->Resize(new_shape, new_type);
+    UpdateViews();
+    return;
+  }
+
+  for (size_t i = 0; i < curr_tensors_size_; i++) {
+    tensors_[i]->Resize(new_shape[i], new_type);
+  }
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::set_type(const TypeInfo &new_type) {
+  type_ = new_type;
+  tl_->set_type(new_type);
+  for (auto t : tensors_) {
+    t->set_type(new_type);
+  }
+  if (state_ == State::contiguous) {
+    UpdateViews();
+  }
+}
+
+
+template <typename Backend>
+const TypeInfo &TensorVector<Backend>::type() const {
+  if (state_ == State::contiguous) {
+    return tl_->type();
+  }
+  if (curr_tensors_size_ == 0) {
+    return type_;
+  }
+  for (size_t i = 1; i < curr_tensors_size_; i++) {
+    assert(tensors_[0]->type() == tensors_[i]->type());
+  }
+  return tensors_[0]->type();
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::SetLayout(const TensorLayout &layout) {
+  if (state_ == State::noncontiguous) {
+    DALI_ENFORCE(!tensors_.empty(), "Layout cannot be set uniformly for empty batch");
+  }
+  tl_->SetLayout(layout);
+  for (auto t : tensors_) {
+    t->SetLayout(layout);
+  }
+}
+
+
+template <typename Backend>
+TensorLayout TensorVector<Backend>::GetLayout() const {
+  if (state_ == State::contiguous) {
+    auto layout = tl_->GetLayout();
+    if (!layout.empty()) return layout;
+  }
+  if (curr_tensors_size_ > 0) {
+    auto layout = tensors_[0]->GetLayout();
+    for (size_t i = 1; i < curr_tensors_size_; i++) assert(layout == tensors_[i]->GetLayout());
+    return layout;
+  }
+  return {};
+}
+
+
+template <typename Backend>
+const DALIMeta &TensorVector<Backend>::GetMeta(int idx) const {
+  assert(static_cast<size_t>(idx) < curr_tensors_size_);
+  return tensors_[idx]->GetMeta();
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::SetMeta(int idx, const DALIMeta &meta) {
+  assert(static_cast<size_t>(idx) < curr_tensors_size_);
+  tensors_[idx]->SetMeta(meta);
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::set_pinned(bool pinned) {
+  // Store the value, in case we pin empty vector and later call Resize
+  pinned_ = pinned;
+  tl_->set_pinned(pinned);
+  for (auto &t : tensors_) {
+    t->set_pinned(pinned);
+  }
+}
+
+
+template <typename Backend>
+bool TensorVector<Backend>::is_pinned() const {
+  if (state_ == State::contiguous) {
+    return tl_->is_pinned();
+  }
+  if (curr_tensors_size_ == 0) {
+    return pinned_;
+  }
+  for (size_t i = 1; i < curr_tensors_size_; i++) {
+    assert(tensors_[i]->is_pinned() == tensors_[0]->is_pinned());
+  }
+  return tensors_[0]->is_pinned();
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::reserve(size_t total_bytes) {
```

Comment:
That's buggy.
1. If previous state was non-contiguous, then we should probably clear `tensors_` (I think that's the only place we should do that).
2. After `tl_->reserve(total_bytes)` there should be a call to UpdateViews.

Reply:
Done